pkg/codegen/importer/hcl2.go

```diff
@@ -365,11 +366,222 @@ func generatePropertyValue(property *schema.Property, value resource.PropertyVal
 	return generateValue(property.Type, value)
 }
 
+// valueStructurallyTypedAs returns true if the given value is structurally typed as the given schema type.
+func valueStructurallyTypedAs(value resource.PropertyValue, schemaType schema.Type) bool {
+	switch {
```

Comment:
We can hoist the union check here, since we need to repeat it for every value anyway:
```suggestion
	if union, ok := schemaType.(*schema.Union); ok {
		schemaType = reduceUnion(union, value)
	}
	switch
```
The reduction is mutually recursive since we want to handle nested discriminated unions.

This lets us remove the `case *schema.UnionType` checks in each branch of the `case value.Is*()`.